# MAEBin: Metagenome binning method integrating Hi-C sequencing and sequence features

MAEBin is a metagenome binning method integrating Hi-C sequencing and sequence features

## Installation

Download the package by 

```
git clone https://github.com/Liudd23/MAEBin.git
cd MAEBin
```

## Required Dependencies

* networkx==3.2.1
* numpy==1.26.3
* torch==2.4.0
* scikit-learn==1.5.2
* dgl==2.4.0
* ogb==1.3.6
* [scipy==1.13.1]
* tensorboardX==2.6.2.2

### An easiler way to install

*Note*: we suggest you to install all the package using conda (both miniconda and [Anaconda](https://anaconda.org/) are ok)

After cloning this respository, you can use anaconda to install the **PHPGATv2.yaml**. The command is: 

```
conda env create -f environment.yml -n MAEBin
```

### Prepare  data

We require the `contig.fa` file, which contains the assembled contig sequences, as well as the `bwa` alignment file (`bwa_output.bam`) generated by aligning sequencing reads to the contigs using the BWA tool. These files are essential for downstream analyses and should be prepared in advance following standard assembly and alignment workflows.

## Usage

    python MAEBin.py [-h] --graphfile GRAPHFILE --embeddingfile EMBEDDINGFILE

### Options

    optional arguments:
      -h, --help            show this help message and exit
      --seeds SEEDS [SEEDS ...]
      --dataset DATASET
      --device DEVICE
      --max_epoch MAX_EPOCH                         number of training epochs
      --warmup_steps WARMUP_STEPS                   number of warmup steps
      --num_heads NUM_HEADS                         number of hidden attention heads
      --num_out_heads NUM_OUT_HEADS                 number of output attention heads
      --num_layers NUM_LAYERS                       number of hidden layers
      --num_hidden NUM_HIDDEN                       number of hidden units
      --residual                                    use residual connection
      --in_drop IN_DROP                             input feature dropout
      --attn_drop ATTN_DROP                         attention dropout
      --norm NORM                                   if norm or not
      --lr LR                                       learning rate
      --weight_decay WEIGHT_DECAY                   weight decay
      --negative_slope NEGATIVE_SLOPE               the negative slope of leaky relu for GAT
      --activation ACTIVATION                       activation
      --mask_rate MASK_RATE                         mask ratio
      --drop_edge_rate DROP_EDGE_RATE               drop edge rate
      --replace_rate REPLACE_RATE                   replace rate
      --encoder ENCODER                             choose encoder
      --decoder DECODER                             choose deconde
      --loss_fn LOSS_FN                             choose loss
      --alpha_l ALPHA_L                            `pow`coefficient for `sce` loss
      --optimizer OPTIMIZER                         choose optimizer
      --deg4feat                                    use node degree as input feature    
      --graphfile GRAPHFILE                         path to Graph file
      --embeddingfile EMBEDDINGFILE                 path to the embedding file

### Example

```
python main_transductive.py \
    --device $device \
    --dataset $dataset \
    --mask_rate 0.7 \
    --encoder "gat" \
    --decoder "gat" \
    --in_drop 0.2 \
    --attn_drop 0.1 \
    --num_layers 1 \
    --num_hidden 512 \
    --num_heads 2 \
    --graphfile GRAPHFILE \ 
    --embeddingfile EMBEDDINGFILE \
```



We also provide a convenient shell script to run directly. You only need to provide the `.fa` file for the contigs and the aligned file.

### Shell script version

```
bash Run_MAEBin.sh [-h] [--contigs CONTIGS] [--bwa-file BWA_FILE] [--output OUTPUT]
```

### Option

```
-h, --help
Show this help message and exit.

--contigs
The path to the contig.fa file containing the assembled contig sequences. This file is required for subsequent steps.

--bwa-file
The path to the bwa alignment file (bwa_output.bam) generated by aligning sequencing reads to the contigs. Ensure the file is prepared using BWA.

--output
The directory where the output files will be saved. Default: ./output.
```

Example

```
# Check help
bash Run_MAEBin.sh -h

# Run the script with contigs, bwa file, and output directory
bash Run_MAEBin.sh --contigs /path/to/contig.fa --bwa-file /path/to/bwa_output.bam --output /path/to/output_dir
```



## Reference

```
@inproceedings{hou2022graphmae,
  title={GraphMAE: Self-Supervised Masked Graph Autoencoders},
  author={Hou, Zhenyu and Liu, Xiao and Cen, Yukuo and Dong, Yuxiao and Yang, Hongxia and Wang, Chunjie and Tang, Jie},
  booktitle={Proceedings of the 28th ACM SIGKDD Conference on Knowledge Discovery and Data Mining},
  pages={594--604},
  year={2022}
}
@article{comebin,
  title={Effective binning of metagenomic contigs using contrastive multi-view representation learning},
  author={Wang, Ziye and You, Ronghui and Han, Haitao and Liu, Wei and Sun, Fengzhu and Zhu, Shanfeng},
  journal={Nature Communications},
  volume={15},
  number={1},
  pages={585},
  year={2024},
  publisher={Nature Publishing Group UK London}
}
@article{metacc,
  title={MetaCC allows scalable and integrative analyses of both long-read and short-read metagenomic Hi-C data},
  author={Du, Yuxuan and Sun, Fengzhu},
  journal={Nature Communications},
  volume={14},
  number={1},
  pages={6231},
  year={2023},
  publisher={Nature Publishing Group UK London}
}
```

## Thank you for using MAEBinÔºÅ
